keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+one is available_, but using other topology hints or ordinary
+`Cluster` traffic policy when not.
+
+[Topology Aware Hints feature]: ../2433-topology-aware-hints/
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+```
+<<[UNRESOLVED deprecate-iTP]>>
+
+- Deprecate `internalTrafficPolicy`? It's clear that the DNS use case
+  given in the Internal Traffic Policy KEP is not actually a good use
+  case for Internal Traffic Policy, because no one wants the behavior
+  of "I'd rather have DNS requests get dropped than have them go to
+  another node". But without the DNS use case, it's not clear that
+  there's really a strong argument for Internal Traffic Policy at all.
+
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Providing any improvements over `externalTrafficPolicy: Local` for
+  LoadBalancer services. Consensus is that [Proxy Terminating
+  Endpoints] should solve the problems that made
+  `externalTrafficPolicy: Local` unreliable for some cases.
+
+- Attempting to improve the endpoint selection algorithm for services
+  with a number of endpoints that is not (roughly) an integer multiple
+  of the number of nodes. If you have a service with 15 endpoints in a
+  cluster with 10 nodes, then the service could potentially be
+  optimized by making clients more likely to select a local endpoint,
+  but they can't _exclusively_ select local endpoints or the endpoints
+  on nodes with only a single endpoint would presumably get
+  overloaded. But this KEP does not attempt to figure out any new
+  behavior there.
+
+[Proxy Terminating Endpoints]: ../1669-graceful-termination-local-external-traffic-policy/
+
+## Proposal
+
+### User Stories
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
```

Comment:
I know we've discussed this before, but I can't remember the background. What was the rationale for this approach over `iTP == PreferLocal`?

Reply:
> you can't put 10 DNS endpoints on 1 node and none anywhere else and expect it to work well as a PreferLocal service.

Well, akshually... why not?  If you assume 10 DNS replicas is sufficient for the whole cluster, you get 1 node which gets fast local access (which is would anyway) and N-1 nodes that get slower remote access (which they would anyway).  PreferLocal == Cluster in that case.

I think the only really problematic scenario is when a node has at least 1 replica, but those local replicas are insufficient to satisfy the node's pods.  E.g. a 10 node cluster has 30 replicas of some service.  1 node gets 1 replica, 2 nodes get 4 replicas, and 7 nodes get 3 replicas.  ON AVERAGE every node has 3, but that node with 1 replica could easily get swamped, while never considering its neighbors who have 3 or 4 replicas.  So there IS still a problem, but I think it's different than you characterized